### PR TITLE
Housekeeping changes

### PR DESCRIPTION
**Pull request by Quafadas**
Upgrades scala 2.13 to 2.13.7

Changes ammonite to 2.4.1

Adds dependancy to os-lib, removes ammonite ops read calls, replace with os lib read

### CODE REVIEW

File: project/deps.sc [hunk 1]
```diff
@@ -62,12 +62,13 @@ object Deps {
...
+  def os_lib                   = ivy"com.lihaoyi::os-lib:0.8.0"
 }
 
 object ScalaVersions {
-  def scala3   = "3.0.1"
-  def cross2_3Version = "2.13.6"
-  def scala213 = "2.13.6"
+  def scala3   = "3.0.2"
+  def cross2_3Version = "2.13.7"
+  def scala213 = "2.13.7"
```
Comment (sideeffffect) [inline project/deps.sc:71]: Besides changing `deps.sc`, you probably need to modify GitHub Actions too https://github.com/almond-sh/almond/blob/master/.github/workflows/ci.yml#L49

File: project/deps.sc [hunk 2]
```diff
@@ -62,12 +62,13 @@ object Deps {
...
   def sparkSql                 = ivy"org.apache.spark::spark-sql:2.4.0"
   def utest                    = ivy"com.lihaoyi::utest:0.7.10"
+  def os_lib                   = ivy"com.lihaoyi::os-lib:0.8.0"
 }
 
 object ScalaVersions {
-  def scala3   = "3.0.1"
-  def cross2_3Version = "2.13.6"
-  def scala213 = "2.13.6"
+  def scala3   = "3.0.2"
```
Comment (sideeffffect) [inline project/deps.sc:69]: There's a 3.1.x version already, btw.

Comment (Quafadas) [reply project/deps.sc:67]: I had a look, as far as I could tell, there is not yet an ammonite version that is published against 3.1.0? It looks like that would get messier than I need... so I will pass on attempting that and wait for the upstream part!

Thankyou very much for the note though :-).

File: project/deps.sc (hunk 1 above)
Comment (Quafadas) [reply project/deps.sc:69]: Oooh - thankyou for your help on my two random pull requests... I think these changes have solved the actual compatibility issues... now apparently the website grumbles :-)!

**Comment by sideeffffect** [comment]
```
Exception in thread "main" java.lang.NoSuchMethodError: pprint.TPrint$.make(Lscala/Function1;)Lpprint/TPrint;
	at mdoc.internal.cli.Settings$.<clinit>(Settings.scala:303)
```
`TPrint.make` seemed to be available in PPrint 0.6.x, but is gone in 0.7.x
https://javadoc.io/static/com.lihaoyi/pprint_2.13/0.6.6/pprint/TPrint$.html#make[T](f:pprint.TPrintColors=%3EString):pprint.TPrint[T]
https://javadoc.io/doc/com.lihaoyi/pprint_2.13/latest/pprint/TPrint$.html

Perhaps we need to upgrade Mill's mdoc plugin or Mill itself? :thinking:

File: project/deps.sc [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import mill._, scalalib._
 
 object Versions {
-  def ammonite      = "2.4.0-14-4824b429"
+  def ammonite      = "2.5.0"
```
Comment (sideeffffect) [inline project/deps.sc:4]: Ammonite 2.5.x depends on PPrint 0.7.x
```
> coursier resolve com.lihaoyi:::ammonite-compiler:2.5.0
...
com.lihaoyi:pprint_2.13:0.7.1:default
```
but mdoc depends on 0.6.x
https://github.com/scalameta/mdoc/blob/v2.2.24/build.sbt#L123

I think we need to stick to older version of Ammonite that contains a compatible version of PPrint: `2.4.1`
https://github.com/com-lihaoyi/Ammonite/blob/2.4.1/build.sc#L96

Would you mind creating a ticket over at mdoc to upgrade to newer version of PPrint? (Could you also, please, mention me?)

**Review by sideeffffect — CHANGES_REQUESTED** [review]

File: project/deps.sc (hunk 3 above)
Comment (Quafadas) [reply project/deps.sc:4]: I opened an issue over at doc. I also had a look at it and concluded that I would get far with attempting that change myelf, sadly. As requested, you got an  @ in the issue. 

Thanks again for your help.

**Comment by sideeffffect** [comment]
Nice, thanks @Quafadas ! Do you think you could verify that all would fork with Ammonite `2.4.1`?

**Comment by Quafadas** [comment]
Apparently, it does!

**Comment by sideeffffect** [comment]
Awesome!

**Comment by Quafadas** [comment]
:-)... It is doubtful these PR's would have "gone green" without your time - many thanks and it was much appreciated.

**Comment by sideeffffect** [comment]
Thanks for helping out with Almond :wink: 
This is a very important project for the Scala community :muscle:

**Comment by Quafadas** [comment]
Hmmm... embarrassingly, I attempted to start a scala 3 kernel, it compiles, but jupyter appears to fall over when attempting to actually use it ... 

```
./mill jupyter 3.0.2
```

```
java.lang.NoClassDefFoundError: ammonite/compiler/Preprocessor
  java.base/java.lang.ClassLoader.defineClass1(Native Method)
  java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
  java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
  java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524)
  java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427)
  java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421)
  java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
  java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420)
  java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
  java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
  almond.amm.AmmInterpreter$$anon$2.<init>(AmmInterpreter.scala:124)
  almond.amm.AmmInterpreter$.apply(AmmInterpreter.scala:135)
  almond.ScalaInterpreter.ammInterp(ScalaInterpreter.scala:106)
  almond.ScalaKernel$$anon$3.run(ScalaKernel.scala:145)
java.lang.ClassNotFoundException: ammonite.compiler.Preprocessor
  java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:445)
  java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
  java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
  java.base/java.lang.ClassLoader.defineClass1(Native Method)
  java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
  java.base/java.security.SecureClassLoader.defineClass(SecureClassLoader.java:150)
  java.base/java.net.URLClassLoader.defineClass(URLClassLoader.java:524)
  java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:427)
  java.base/java.net.URLClassLoader$1.run(URLClassLoader.java:421)
  java.base/java.security.AccessController.doPrivileged(AccessController.java:712)
  java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:420)
  java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:587)
  java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
  almond.amm.AmmInterpreter$$anon$2.<init>(AmmInterpreter.scala:124)
  almond.amm.AmmInterpreter$.apply(AmmInterpreter.scala:135)
  almond.ScalaInterpreter.ammInterp(ScalaInterpreter.scala:106)
  almond.ScalaKernel$$anon$3.run(ScalaKernel.scala:145)
```

So embarrassingly, it would appear to compile, but not "work" ...

**Comment by sideeffffect** [comment]
Hmm... are you sure that you have the jar https://repo1.maven.org/maven2/com/lihaoyi/ammonite-compiler_3.0.2/2.4.1/ammonite-compiler_3.0.2-2.4.1.jar on the classpath?

Because the class seems to be available:
https://javadoc.io/doc/com.lihaoyi/ammonite-compiler_3.0.2/2.4.1/api/ammonite/compiler/Preprocessor.html

File: project/deps.sc [hunk 4]
```diff
@@ -31,9 +31,7 @@ implicit class DepOps(private val dep: Dep) {
 }
 
 object Deps {
-  def ammoniteCompiler(sv: String)             =
-    if (sv.startsWith("2.")) ivy"com.lihaoyi:::ammonite-compiler:${Versions.ammonite}"
-    else ivy"com.lihaoyi:ammonite-compiler_$sv:${Versions.ammonite}"
+  def ammoniteCompiler(sv: String) = ivy"com.lihaoyi:::ammonite-compiler:${Versions.ammonite}"
```
Comment (sideeffffect) [inline project/deps.sc:34]: How about
```suggestion
  def ammoniteCompiler(sv: String) = ivy"com.lihaoyi:ammonite-compiler_$sv:${Versions.ammonite}"
```
?
Or try reverting this change altogether? :shrug:

File: project/deps.sc [hunk 5]
```diff
@@ -62,17 +60,18 @@ object Deps {
   def slf4jNop                 = ivy"org.slf4j:slf4j-nop:1.8.0-beta4"
   def sparkSql                 = ivy"org.apache.spark::spark-sql:2.4.0"
   def utest                    = ivy"com.lihaoyi::utest:0.7.10"
+  def os_lib                   = ivy"com.lihaoyi::os-lib:0.8.0"
 }
 
 object ScalaVersions {
-  def scala3   = "3.0.1"
+  def scala3   = "3.0.2"
   def cross2_3Version = "2.13.6"
```
Comment (sideeffffect) [inline project/deps.sc:68]:
```suggestion
  def cross2_3Version = "2.13.7"
```

**Comment by Quafadas** [comment]
Screenshot 2022-01-11 at 21 29 59

So I'm pretty sure there's something scala 3 in there, even if it's telling me it's scala 2.13.6... 

It would appear that ammonite doesn't support defining extension methods - fair enough, as they are somewhat new I believe. 
https://github.com/com-lihaoyi/Ammonite/issues/1238

That was a more time consuming odyssey than expected - I hope it saves someone else at least some time. Thanks again for the help.

File: project/deps.sc [hunk 6]
```diff
@@ -31,9 +31,10 @@ implicit class DepOps(private val dep: Dep) {
 }
 
 object Deps {
-  def ammoniteCompiler(sv: String)             =
-    if (sv.startsWith("2.")) ivy"com.lihaoyi:::ammonite-compiler:${Versions.ammonite}"
-    else ivy"com.lihaoyi:ammonite-compiler_$sv:${Versions.ammonite}"
+  
+  def ammoniteCompiler(sv: String) = {
+      if (sv.startsWith("2.")) ivy"com.lihaoyi:::ammonite-compiler:${Versions.ammonite}"
+      else ivy"com.lihaoyi:ammonite-compiler_$sv:${Versions.ammonite}"}
```
Comment (sideeffffect) [inline project/deps.sc:37]: You probably want to revert the changes here :wink:

**Review by sideeffffect — CHANGES_REQUESTED** [review]

File: project/deps.sc (hunk 6 above)
Comment (Quafadas) [reply project/deps.sc:37]: I think that it was trying to move this artefact to the "native" scala 3 which caused the trouble. 

I suspect the 3 ammonite artefacts are quite closely coupled and I should not have attempted to tamper with them. 

I've compared these lines with the existing master branch, and as far as I can tell, they are the same... from my experiments, this PR now behaves as expected, with the confusing behaviour of reporting a 2.13 version string. 

Maybe this PR helps the maintainers, maybe not - at least the intent was the right one!  I think I'll down tools here though.

Thankyou again for time and suggestions, I learned a lot and it was appreciated.

Comment (sideeffffect) [reply project/deps.sc:37]: > I've compared these lines with the existing master branch, and as far as I can tell, they are the same

They're not. Just have a look here https://github.com/almond-sh/almond/pull/902/files
:wink:

**Comment by sideeffffect** [comment]
It's trying to download cs version `2.1.0-M5-24-g678b31710` but that isn't found/doesn't exist... :thinking:

**Comment by Quafadas** [comment]
I think I broke this beyond repair trying to fix the website thing. But learned enough to start from the last sensible point and make CI go green again... 

See #930
